[Bots] Line of Sight and Mez optimizations and cleanup

nytmyr opened the pull request:
# Description

- Renames `Map:CheckForLoSCheat` to `Map:CheckForDoorLoSCheat` to better reflect what it does.
- Renames `Map:RangeCheckForLoSCheat` to `Map:RangeCheckForDoorLoSCheat` to better reflect what it does.
- Adds the rule `Pets:PetsRequireLoS` to determine whether or not commanded pet attacks require an additional layer of LoS checks for edge-cases.
- Adds the rule `Bots:BotsRequireLoS` to determine whether or not bots require LoS to `^attack`, `^pull` and `^precombat`.
- Adds the rule `Map:ZonesToCheckDoorCheat` to control which, if any, zones will be checked.
- Corrects, removes and adds LoS checks where necessary and eliminated unnecessary duplicate checks.
- Improves door checking logic for locked or triggered doors that could be blocking LoS.
- Cleans up false positives for door cheat checks.
- Adds `drawbox` option to `#door` command. This will spawn points at the center and each corner of the door's "box". It will also spawn points at your and your target's location.
- Improves Mez and AE Mez logic
- Adds more details to the rule `Bots:EpicPetSpellName`
- Door logic can be further improved upon at a later date to account for different opening types if needed.
 - This is door cheat is really only intended to be used for old zones where the client vs server see some doors differently and the server will pass line of sight checks through old doors for certain commanded things.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Testing

- Many LoS scenario checks involving pets, bots, combat, casting and AE logic.

Clients tested: RoF2

# Checklist

- [x] I have tested my changes
- [x] I have performed a self-review of my code. Ensuring variables, functions and methods are named in a human-readable way, comments are added only where naming of variables, functions and methods can't give enough context.
- [x] I own the changes of my code and take responsibility for the potential issues that occur

nytmyr committed:
[Bots] Line of Sight and Mez optimizations and cleanup

- Renames `Map:CheckForLoSCheat` to `Map:CheckForDoorLoSCheat` to better reflect what it does.
- Renames `Map:RangeCheckForLoSCheat` to `Map:RangeCheckForDoorLoSCheat` to better reflect what it does.
- Adds the rule `Pets:PetsRequireLoS` to determine whether or not commanded pet attacks require an addition layer of LoS checks for edge-cases.
- Adds the rule `Bots:BotsRequireLoS` to determine whether or not bots require LoS to `^attack`, `^pull` and `^precombat`.
- Adds the rule `Map:ZonesToCheckDoorCheat` to control what if any zones will be checked..
- Corrects, removes and adds LoS checks where necessary.
- Improves door checking logic for locked or triggered doors that could be blocking LoS.
- Cleans up false positives for door cheat checks.
- Adds `drawbox` option to `#door` command. This will spawn points at the center and each corner of the door's "box". It will also spawn points at your and your target's location.
- Improves Mez and AE Mez logic
- Adds more details to the rule `Bots:EpicPetSpellName`

nytmyr committed:
[Bots] Line of Sight and Mez optimizations and cleanup

- Renames `Map:CheckForLoSCheat` to `Map:CheckForDoorLoSCheat` to better reflect what it does.
- Renames `Map:RangeCheckForLoSCheat` to `Map:RangeCheckForDoorLoSCheat` to better reflect what it does.
- Adds the rule `Pets:PetsRequireLoS` to determine whether or not commanded pet attacks require an addition layer of LoS checks for edge-cases.
- Adds the rule `Bots:BotsRequireLoS` to determine whether or not bots require LoS to `^attack`, `^pull` and `^precombat`.
- Adds the rule `Map:ZonesToCheckDoorCheat` to control what if any zones will be checked..
- Corrects, removes and adds LoS checks where necessary.
- Improves door checking logic for locked or triggered doors that could be blocking LoS.
- Cleans up false positives for door cheat checks.
- Adds `drawbox` option to `#door` command. This will spawn points at the center and each corner of the door's "box". It will also spawn points at your and your target's location.
- Improves Mez and AE Mez logic
- Adds more details to the rule `Bots:EpicPetSpellName`

nytmyr commented:
> Renaming Rules can have adverse effects across the community as the cleanup will remove the old rule and what ever it was set for and set the new default. Just be aware of what can happen by doing so.

Thought about that with this one. Figured there's likely nobody actually using this rule ATM being freshly in the release and defaulted to false. Even if they were, it wasn't a reliable check in its previous form and had false positives. I can switch it to false by default or add a DB update to check and insert the rule based off current settings. Didn't feel the need to add a manifest update for this edge case rule but can if desired.

nytmyr commented:
Marked as draft till decision on rule change is made

nytmyr committed:
[Hotfix] Fix cursor load on zone (#4704)

* [Hotfix] Fix cursor load on zone

- Corrects an issue where the item on your cursor would become invisible upon trying to place it anywhere in your inventory.

* Remove general to prevent duplicate loads

Kinglykrab committed:
[Quest API] Add GetSpawn() to Perl and Lua (#4702)

* [Quest API] Add GetSpawn() to Perl and Lua

* Push

* Update npc.h

Kinglykrab committed:
[Bug Fix] Fix Lua Zone ID Exports (#4700)

nytmyr committed:
[Bots] Prevent medding in combat if any mob has bot targeted (#4707)

- Previously this only checked the bots target to med. It will now check the hatelist of the bot to ensure no mobs have the bot targeted before medding.

nytmyr committed:
[Bots] Improve positioning (#4709)

- Prevent bots not set to ^behindmob from finding positions behind their target.
- Backs up a bot if they're too close and have target reflection, not just taunting bots.
- Backs up bots if the target is rooting and they are not taunting to the appropriate position.
- This will particularly help with casters running around.

nytmyr committed:
[Bots] Add checks to ensure bots and pets do not engage on ^pull (#4708)

zimp-wow committed:
[Bug Fix] Fix bad Mob reference in QuestManager::resumetimer() (#4710)

Akkadius committed:
[Release] 23.0.2 (#4711)

Akkadius committed:
[Tasks] Extend IsTaskCompleted to also be aware of shared task completion (#4714)

* [Tasks] Extend IsTaskCompleted to also be aware of shared task completion

* Fix my stupidity

* Update client.h

Akkadius committed:
[Hotfix] Clear m_completed_shared_tasks before reloading

Kinglykrab committed:
[Bug Fix] Fix Issue with Suffixes/Prefixes (#4723)

Kinglykrab committed:
[Bug Fix] Fix AA Reset Error Message (#4720)

* [Bug Fix] Fix AA Reset Error Message

* Update client.cpp

* Update client.cpp

Kinglykrab committed:
[Quest API] Add Key Ring Methods to Perl and Lua (#4719)

Akkadius committed:
[Quest API] Implement eq.handin() and quest::handin() (#4718)

* [Quest API] Implement eq.handin() and quest::handin()

* Fix MQ using new API style

Akkadius committed:
[Zone] Implement Zone State Saving on Shutdown (#4715)

* Save spawns

* Update base_zone_state_spawns_repository.h

* Zone state save work

* Code cleanup

* More cleanup

* Database migration

* Update database_update_manifest.cpp

* Revert decay at storage model

* Code cleanup

* More cleanup

* More cleanup

* More cleanup

* Entity variables

* Add entity variables to the schema

* Post rebase

* Checkpoint

* Serialize / deserialize buffs

* Current hp / mana / end save / load

* Save / load current_waypoint

* Add zone spawn protection

* Finishing touches

* Cleanup

* Update zone_save_state.cpp

* Cleanup

* Update zone_save_state.cpp

* Update npc.cpp

* Update npc.cpp

* More

* Update perl_npc.cpp

* Update zone_loot.cpp

MortimerGreenwald committed:
[Bug Fix] Fix Trading Items to Bot Pets (#4721)

Small change using IsPetOwnerOfClientBot() instead of just IsPetOwnerClient

Can bot pet take item?

not before, but hopefully now?

Akkadius committed:
[Code Cleanup] More login <-> world code cleanup (#4724)

* More cleanup

* More cleanup

nytmyr committed:
[Bots] Fix unresponsive bots in groups upon group wipe (#4712)

- Bots were failing to recall their group after dying due to the way it was grabbing the group from entity list.

Akkadius commented on zone/botspellsai.cpp:
```diff
@@ -219,7 +216,7 @@ bool Bot::AICastSpell(Mob* tar, uint8 chance, uint16 spell_type, uint16 sub_targ
 	for (const auto& s : bot_spell_list) {
 
-		if (!IsValidSpellAndLoS(s.SpellId, HasLoS())) {
+		if (!IsValidSpellAndLoS(s.SpellId, (!BotRequiresLoSToCast(spell_type, s.SpellId) ? true : HasLoS()))) {
```
This is extremely gnarly to read

`IsValidSpellAndLoS` should just validate a valid spell, not also check LOS

There should be a separate check that handles LOS requirements of a spell which can do ternary stuff inside it and make things a little easier to read. This is duplicated in a few other spots.

catapultam-habeo committed:
[Feature] Add a rule for spells to bypass stacking rules (#4716)

* define rule that allows for spells which always stack

* Update spells.cpp

---------

Co-authored-by: Akkadius <akkadius1@gmail.com>

Akkadius commented on zone/mob.cpp:
```diff
@@ -8726,27 +8740,24 @@ bool Mob::CheckLosCheat(Mob* other) {
 	return true;
 }
 
-bool Mob::CheckLosCheatExempt(Mob* other)
```
This is all commented out and just returns false ? Should we just remove this code ?

catapultam-habeo committed:
[Bug Fix] Refactor ApplyItemBonuses to fix double-counting of ATK and recommended levels not correctly applying (#4713)

* refactor AddItemBonuses

* typo

* simplify some syntax

* fix indents

* Revert "fix indents"

This reverts commit 8e660707a96090105213fcee924d0e492fc44ddd.

* fix indents without blowing up entire file

* Revert "fix indents without blowing up entire file"

This reverts commit 6b21d1bcc96608e1dc620962b426401540cedbbf.

* ok for real though

* Undo spaces in formatting, source uses tabs

* Spaces to tabs take 2

---------

Co-authored-by: Akkadius <akkadius1@gmail.com>

neckkola committed:
[Feature] Evolving items Additions (#4725)

* Implement multi-value for evolving sub_types

- Added ability for evolving sub_types to contain multiple values
- Implemented EvolvingItems::Types::NUMBER_OF_KILLS with level for sub_type

* Repair a timer issue preventing proper evolution of items

* Simplify

* Remove extra level of nesting

* Update client_evolving_items.cpp

---------

Co-authored-by: Akkadius <akkadius1@gmail.com>

Akkadius committed:
[Zone State] Wrap all serialization/deserialization in try/catch (#4726)

Akkadius committed:
[Database] Remove force_interactive from big bag updates (#4727)

Akkadius committed:
[Crash] Fix filesystem crash / exception in DatabaseDumpService::RemoveCredentialsFile() (#4731)

Akkadius committed:
[Loginserver] Minor cleanup (#4729)

Akkadius committed:
[Crash] World CLI validation (#4728)

* [Crash] World CLI validation

* Clean

neckkola committed:
[Fix] Add client packets to questmanager:setguild (#4732)

Akkadius committed:
[Crash] Check for directory existence before traversing in CheckForCompatibleQuestPlugins (#4730)

Akkadius committed:
[Crash] Fix reload concurrency crash when ran from Spire (#4733)

Akkadius committed:
[Crash] Fix large file size crash in File::GetContents for windows (#4735)

Akkadius committed:
[Crash] Validate item in SE_SummonItemIntoBag (#4734)

Akkadius committed:
[Hotfix] Add character_pet_name to player tables schema

Akkadius committed:
[Release] 23.1.0 (#4736)

* [Release] 23.1.0

* Version

* Pet table last minute add in

catapultam-habeo committed:
[Bug Fix] Fix an error causing Endurance Regen to not be applied by items. (#4738)

* fix typo causing endurance regen to not be applied by items

* further correction

zimp-wow committed:
[Bug Fix] Cleanup zone buckets on instance purge. (#4739)

Akkadius committed:
[World] Check if port in use to avoid double booting mistakes (#4740)

* Stuff

* Potentially fix aura crash

* Reload crash fix

* Revert "Reload crash fix"

This reverts commit 96e1e76306059fc2a624e0152eca4beaee49a0f8.

* Fix

* Update entity.cpp

* Update dbcore.cpp

* [World] Check if port in use to avoid double booting mistakes

* Revert "Stuff"

This reverts commit 2162c00eddfd13c7ff9fbe53cada1ea36373880c.

* Revert "Potentially fix aura crash"

This reverts commit 7c242723f40a0d013371c36e205a9917849ee862.

* Revert "Fix"

This reverts commit 8419e284d48bac823c861c480719eb3d3b8290a9.

* Revert "Update entity.cpp"

This reverts commit 8a1f4545a4e1c96a848866fb5b6185c66db09097.

* Revert "Update dbcore.cpp"

This reverts commit f0278d95910ccf9e1667eb0e1a0f2608aa7b6058.

Akkadius committed:
[Crash] Database SetMutex crash fix (#4741)

Akkadius committed:
[Crash] Fix world repop crash (#4742)

Akkadius committed:
[Crash] Fix Aura process crash with bots (#4743)

Akkadius committed:
[Crash] Potential crash fix in scan close mobs (#4744)

Akkadius committed:
[Crash] Fix crash in add loot code path (#4745)

Akkadius committed:
[Release] 23.2.0

Akkadius committed:
[Hotfix] Remove one port check in world

Akkadius committed:
[Hotfix] Fix sigabort crash from invalid JSON

nytmyr committed:
Cleanup

nytmyr committed:
Cleanup

nytmyr commented:
Oh god, this got fucked. Closing and adding a new PR
